### PR TITLE
Improve error handling in flagsaver.

### PR DESCRIPTION
**Pull request by Martin2112**
Exclude saving of "go test" related flags and problematic
log_backtrace_at flag. All other errors related to setting flags are
reported. Add MustRestore() to make tests cleaner.

### Checklist

- [ ] I have updated the [CHANGELOG](CHANGELOG.md).
  - Adjust the draft version number according to [semantic versioning](https://semver.org/) rules.
- [ ] I have updated [documentation](docs/) accordingly.

### CODE REVIEW

**Comment by Martin2112** [comment]
Ping.

File: util/flagsaver/flagsaver.go [hunk 1]
```diff
@@ -42,9 +50,23 @@ func Save() *Stash {
...
+		if !strings.HasPrefix(f.Name, "test.") && f.Name != "log_backtrace_at" {
+			s.flags[f.Name] = f.Value.String()
+		}
 	})
 
 	return &s
 }
+
+// MustRestore calls Restore and exits on failure. It can be used in a defer for
+// tests. If Restore fails then otherwise the flags may be in an arbitrary
```
Comment (RJPercival) [inline util/flagsaver/flagsaver.go]:
```suggestion
// tests. If Restore fails then the flags may be in an arbitrary
```

File: util/flagsaver/flagsaver_test.go [hunk 2]
```diff
@@ -94,8 +94,12 @@ func TestRestore(t *testing.T) {
 		}
 
 		func() {
-			defer Save().Restore()
-			flag.Set(test.flag, test.newValue)
+			defer Save().MustRestore()
+			// If the Set() fails the value won't have been updated but some of the
+			// test cases set the same value so it's safer to have this check.
+			if err := flag.Set(test.flag, test.newValue); err != nil {
+				t.Errorf("%v: flag.Set(%q)=%v", test.desc, test.flag, err)
```
Comment (RJPercival) [inline util/flagsaver/flagsaver_test.go]:
```suggestion
				t.Errorf("%v: flag.Set(%q) = %q, want nil", test.desc, test.flag, err)
```

**Comment by Martin2112** [comment]
We should maybe move this code somewhere that's not part of test coverage?

**Comment by RJPercival** [comment]
Perhaps the testonly directory? I'm not sure there's a use case for it outside of tests.

**Comment by Martin2112** [comment]
Yep I think that's a good idea.

**Pull request merged**